File: packages/ios-uikit/Sources/FarmerChatUIKit/FarmerChat.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif

/// Main entry point for the FarmerChat UIKit SDK.
///
/// Initialize once in `AppDelegate.application(_:didFinishLaunchingWithOptions:)`:
/// ```swift
/// FarmerChat.shared.initialize(config: FarmerChatConfig(apiKey: "your-key"))
/// ```
@MainActor
public final class FarmerChat {

    private static let tag = "FC.FarmerChat"

    public static let shared = FarmerChat()
    private init() {}

    public private(set) var isInitialized = false
    public private(set) var config: FarmerChatConfig?

    internal var apiClient: ApiClient?
    internal var connectivityMonitor: ConnectivityMonitor?
    internal var crashBridge: CrashBridge?
    internal var eventCallback: ((FarmerChatEvent) -> Void)?
    private var generatedSessionId: String?

    // MARK: - Public API

    public func initialize(config: FarmerChatConfig) {
        guard !isInitialized else { return }
        do {
            try setupSDK(config: config)
        } catch {
            print("[\(FarmerChat.tag)] Initialization error: \(error)")
        }
    }

    private func setupSDK(config: FarmerChatConfig) throws {
        self.config = config
        self.generatedSessionId = UUID().uuidString
        self.eventCallback = config.onEvent

        guard let baseURL = URL(string: config.baseUrl),
              let scheme = baseURL.scheme?.lowercased(),
              scheme == "http" || scheme == "https" else {
            print("[\(FarmerChat.tag)] Invalid base URL: \(config.baseUrl)")
            return
        }

        let deviceInfo = "iOS/UIKit/\(UIDevice.current.systemVersion)"
        self.apiClient = ApiClient(
            baseUrl: config.baseUrl,
            sdkApiKey: config.apiKey,
            deviceInfo: deviceInfo,
            timeoutMs: config.requestTimeoutMs
        )

        let monitor = ConnectivityMonitor()
        monitor.start()
        self.connectivityMonitor = monitor

        let bridge = CrashBridge()
        bridge.detect()
        bridge.setCustomKey("sdk_version", value: "0.0.0")
        if let partnerId = config.partnerId { bridge.setCustomKey("partner_id", value: partnerId) }
        self.crashBridge = bridge

        self.isInitialized = true
        print("[\(FarmerChat.tag)] SDK initialized")

        // Kick off guest user initialization in the background
        let deviceId = SdkPreferences.stableDeviceId
        let baseUrl  = config.baseUrl
        Task {
            await self.ensureGuestUser(baseUrl: baseUrl, deviceId: deviceId)
        }
    }

    internal func ensureGuestUser(baseUrl: String, deviceId: String) async {
        await TokenStore.shared.setDeviceId(deviceId)
        let isInitialized = await TokenStore.shared.isInitialized
        guard !isInitialized else { return }
        do {
            let guestClient = GuestAPIClient(baseUrl: baseUrl)
            _ = try await guestClient.initializeUser(deviceId: deviceId)
            print("[\(FarmerChat.tag)] Guest user initialized")
        } catch {
            print("[\(FarmerChat.tag)] Guest user init failed: \(error)")
        }
    }

    public func destroy() {
        connectivityMonitor?.stop()
        connectivityMonitor = nil
        apiClient = nil
        crashBridge = nil
        eventCallback = nil
        config = nil
        generatedSessionId = nil
        isInitialized = false
    }

    #if canImport(UIKit)
    /// Returns a `UINavigationController` hosting the chat UI.
    /// Shows onboarding on first launch; shows chat on subsequent launches.
    public func chatViewController() -> UINavigationController {
        guard isInitialized else {
            print("[\(FarmerChat.tag)] Call initialize(config:) first.")
            let vc = UIViewController()
            vc.view.backgroundColor = .systemBackground
            return UINavigationController(rootViewController: vc)
        }

        let rootVC: UIViewController
        if !SdkPreferences.isOnboardingDone {
            rootVC = OnboardingViewController()
        } else {
            rootVC = ChatViewController()
        }
        let nav = UINavigationController(rootViewController: rootVC)
        nav.modalPresentationStyle = .fullScreen
        return nav
    }

    public func fabView(action: @escaping () -> Void) -> UIView {
        let fab = FarmerChatFAB()
        fab.tapAction = action
        return fab
    }
    #endif

    // MARK: - Internal

    internal func getSessionId() -> String {
        config?.sessionId ?? generatedSessionId ?? ""
    }

    nonisolated internal static func getConfig() -> FarmerChatConfig {
        MainActor.assumeIsolated { shared.config ?? FarmerChatConfig() }
    }
}

File: packages/android-views/src/main/kotlin/org/digitalgreen/farmerchat/views/FarmerChat.kt
package org.digitalgreen.farmerchat.views

import android.content.Context
import android.util.Log
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import org.digitalgreen.farmerchat.views.crash.CrashBridge
import org.digitalgreen.farmerchat.views.network.ApiClient
import org.digitalgreen.farmerchat.views.network.ConnectivityMonitor
import org.digitalgreen.farmerchat.views.network.DeviceInfoProvider
import org.digitalgreen.farmerchat.views.network.GuestApiClient
import org.digitalgreen.farmerchat.views.network.TokenStore

/**
 * Main entry point for the FarmerChat XML Views SDK.
 * Initialize once in your Application.onCreate().
 */
object FarmerChat {

    private const val TAG = "FarmerChat"
    internal const val SDK_VERSION = "0.0.0"

    private var config: FarmerChatConfig? = null
    private var sdkApiKey: String? = null
    private var appContext: Context? = null
    @Volatile private var isInitialized = false
    private var sessionId: String? = null
    internal var apiClient: ApiClient? = null
        private set
    internal var guestApiClient: GuestApiClient? = null
        private set
    internal var connectivityMonitor: ConnectivityMonitor? = null
        private set
    internal var crashBridge: CrashBridge? = null
        private set
    internal var eventCallback: ((FarmerChatEvent) -> Unit)? = null
        private set

    private val sdkScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    /**
     * Initialize the SDK. Call once in Application.onCreate().
     *
     * This method is idempotent — calling it more than once is a no-op.
     * All work is wrapped in try-catch so SDK initialization can never crash the host app.
     *
     * @param context   Application context
     * @param sdkApiKey Partner API key (`fc_live_*` or `fc_test_*`)
     * @param config    Optional SDK configuration
     */
    fun initialize(
        context: Context,
        sdkApiKey: String,
        config: FarmerChatConfig = FarmerChatConfig(),
    ) {
        if (isInitialized) return
        synchronized(this) {
            if (isInitialized) return
            try {
                val appCtx = context.applicationContext
                this.appContext = appCtx
                this.sdkApiKey = sdkApiKey
                this.config = config
                this.sessionId = java.util.UUID.randomUUID().toString()

                val deviceId = DeviceInfoProvider.getStableDeviceId(appCtx)
                TokenStore.setDeviceId(deviceId)
                val deviceInfo = DeviceInfoProvider.buildHeader(appCtx)

                this.guestApiClient = GuestApiClient(config.baseUrl)
                this.apiClient = ApiClient(
                    baseUrl     = config.baseUrl,
                    sdkApiKey   = sdkApiKey,
                    deviceInfo  = deviceInfo,
                    timeoutMs   = config.requestTimeoutMs,
                    aiReadTimeoutMs = config.aiReadTimeoutMs,
                )
                this.connectivityMonitor = ConnectivityMonitor(appCtx).also { it.start() }
                this.crashBridge = CrashBridge().also { it.detect() }
                this.isInitialized = true

                // Background guest token initialization
                sdkScope.launch {
                    try {
                        if (!TokenStore.isInitialized) {
                            guestApiClient?.initializeUser(deviceId)
                        }
                    } catch (e: Exception) {
                        Log.w(TAG, "Background guest init failed: ${e.message}")
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Initialization failed", e)
            }
        }
    }

    /**
     * Check if the SDK has been initialized.
     */
    fun isInitialized(): Boolean = isInitialized

    /**
     * Returns the current SDK configuration, or defaults if not yet initialized.
     */
    fun getConfig(): FarmerChatConfig = config ?: FarmerChatConfig()

    /**
     * Returns the current session ID, or empty string if not yet initialized.
     */
    fun getSessionId(): String = sessionId ?: ""

    /**
     * Returns the application context. For SDK internal use only.
     *
     * @throws IllegalStateException if the SDK has not been initialized.
     */
    internal fun getContext(): Context = requireNotNull(appContext) { "FarmerChat not initialized" }

    /**
     * Set event callback for SDK lifecycle events.
     *
     * @param callback Invoked on the main thread for each [FarmerChatEvent].
     */
    fun setEventCallback(callback: (FarmerChatEvent) -> Unit) {
        this.eventCallback = callback
    }

    /**
     * Destroy the SDK and release all resources.
     * After calling this, [initialize] may be called again.
     */
    fun destroy() {
        synchronized(this) {
            try {
                connectivityMonitor?.stop()
                apiClient = null
                connectivityMonitor = null
                crashBridge = null
                config = null
                apiKey = null
                appContext = null
                sessionId = null
                eventCallback = null
                isInitialized = false
            } catch (e: Exception) {
                Log.e(TAG, "Destroy failed", e)
            }
        }
    }
}
